cmake: refactor `zephyr_get_*_for_lang` functions

pillo79 opened the pull request:
**Not for 4.3.**

This PR unifies the implementation of the various `zephyr_get_*_for_lang` functions and their string variants with a generic `zephyr_get_property`, so that the same feature set is available to all wrappers. It also adds several new options that expand the possible use cases:

  - `AS_STRING`: Return the result as a string rather than a list. Equivalent to specifying `DELIMITER " "`.
  - `DELIMITER <delim>`: Specify the output delimiter to use for the list entries. Defaults to `$<SEMICOLON>`.
  - `GENEX`: Use generator expressions to get the property entries at build time, for targets that are not yet fully configured. Cannot be used with `LANG`. Also, CMake will complain if the requested property includes `$<COMPILE_OPTIONS:...>` expressions.
  - `LANG <C|CXX|ASM>`: When set, the property value is filtered for `$<COMPILE_LANGUAGE:lang>` entries, and only those matching the given language are kept. Cannot be used with `GENEX`.
  - `PREFIX <prefix>`: Specify a prefix to use for each entry in the output list.
  - `STRIP_PREFIX`: Omit the prefix string, even if given by `PREFIX`.
  - `TARGET <target>`: Specify the target to get the property from. Defaults to `zephyr_interface`.

This work was done while experimenting with alternatives to #98348. It is no longer useful for the EDK, but I think it is still worth considering for merge.

pillo79 committed:
cmake: (refact) add zephyr_get_build_property

Refactored the various 'zephyr_get_*_for_lang*' functions into a single
'zephyr_get_build_property' function that takes the property to get and
the specific output formats as arguments. This avoids differences
between each function, reduces code duplication, and makes it easier to
support additional features in the future.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
cmake: (refact) drop redundant code

The 'convert_list_of_flags_to_string_of_flags' function was added back
in 2017 with commit 89516fbc25f0 at the time language support was added
to the CMake build system. The same commit also added all users of that
function; they were directly passed the result of get_property() calls,
so a simple replace was sufficient.

This code was later expanded with the use of generator expressions to
manage delimiters and prefixes, which required replacing the ';'
separator with '$<SEMICOLON>' to properly preserve the list entries.

This made the conversion function invalid, since no ';' can be found in
any 'zephyr_get_*' output, so the conversion function was effectively a
no-op; setting a space as a delimiter is already sufficient to get the
desired result since circa 2020.

Remove this function and 'get_property_and_add_prefix', a macro that was
made redundant by the same value-to-genex conversion.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -164,11 +164,56 @@ function(zephyr_ld_options)
+#  - STRIP_PREFIX: Omit the compiler flag prefix (-I, -D, etc.)
+#  - DELIMITER <delimiter>: Specify the output delimiter to use
+
+function(zephyr_get_build_prop_for_lang prop lang i)
```
I think cleaning up the current use of `zephyr_get_*_for_lang..` is valuable, but if we anyway change current style, then perhaps align closer to CMake.

Like we have `zephyr_string` and `zephyr_get`, then perhaps a
```
zephyr_get_property(<out-var> PROPERTY <prop> LANG <C|CXX|ASM>)
```
alternative `zephyr_get_build_property()`.

Optional, the `_as_string` could even be an extra flag to the function instead of dedicated function.

pillo79 replied:
I dropped `AS_STRING` once I understood it could be done with existing options. Looks a lot cleaner with it though, so I'll re-add that.

tejlmand reviewed (commented):
I like the direction, but couple of small adjustments proposed.

pillo79 reviewed (commented):
Thanks, will update soon with comments.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -183,109 +259,67 @@ endfunction()
+function(zephyr_get_system_include_directories_for_lang_as_string lang i)
+  zephyr_get_property(result_output_list
+                      PROPERTY INTERFACE_SYSTEM_INCLUDE_DIRECTORIES
+                      PREFIX "-isystem" DELIMITER " "
```
`AS_STRING` ?
```suggestion
                      PREFIX "-isystem" AS_STRING
```

tejlmand reviewed (commented):
Thanks for the update :+1: 

Mostly some minor observations and clarification questions.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -164,13 +164,89 @@ function(zephyr_ld_options)
+    string(REPLACE ";" "$<SEMICOLON>" genexp_output_list "${output_list}")
+  endif()
+
+  set(result_output_list "${maybe_prefix}$<JOIN:${genexp_output_list},${args_DELIMITER}${maybe_prefix}>")
```
I think this is the expected behavior? If the user asks for space as a delimiter, having the char in the`${args_DELIMITER}` variable will properly expand the genex to a single-space-separated string.

The linked section of the docs mention that the resulting string becomes ambiguous if there were spaces in the data (you may either have too many arguments, or a single quote-enclosed one), but to avoid that... you should use a proper list in the first place (which is returned without `AS_STRING`).

pillo79 reviewed (commented):
Thanks, I understand the concerns, here are my thoughts on them. Let me know how to address the prefix thing.

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -183,109 +259,67 @@ endfunction()
-  set(result_output_list "-D$<JOIN:${genexp_output_list},${args_DELIMITER}-D>")
+  zephyr_get_property(result_output_list
+                      PROPERTY INTERFACE_COMPILE_DEFINITIONS
+                      PREFIX "-D"
```
>  but I disliked it for a number of reasons:

yes, seeing it from the function implementation then the `if`'s is really not nice.
I also understand from the concern on the build specific part of the function.

However, when seeing it from the caller, that is if we want to get rid of `zephyr_get_*_for_lang` functions, then requesting the caller to remember `PREFIX -D` when using `INTERFACE_COMPILE_DEFINITIONS` is not nice.
`-D` ought to be implied by the property name (when seeing things from the caller).

> Let me know if you still prefer to incorporate the prefixes in the generic function.

Depends on what you believe the long term support for existing `zephyr_get_*_for_lang` functions is.

> If so, I would suggest to call that zephyr_get_build_property as it's doing build-specific stuff as well.

That might be a better name. Especially as the function description already states:
https://github.com/zephyrproject-rtos/zephyr/blob/2fa423179490dbc4bec3f47581368089b4c8fb1c/cmake/modules/extensions.cmake#L167-L168

so it is target related, and targets are build specific, as well as `LANG` is very build oriented.

pillo79 committed:
cmake: zephyr_get_build_property: implement additional options

Extended the zephyr_get_build_property() CMake function with several
convenience options:

 - GENEX:
      Uses generator expressions to get the property entries at build
      time, for targets that are not yet fully configured.
      Cannot be used with LANG. Also, CMake will complain if the
      requested property includes $<COMPILE_OPTIONS:...> expressions.

 - TARGET <target>:
      Specifies the target to get the property from. Defaults to
      "zephyr_interface" as before.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
@tejlmand PTAL, pity to leave this to rot.
